File: scala/src/P04PartialApplication.scala
object P04PartialApplication {
  def main(args: Array[String]): Unit = {

    val partial1 = f(1)
    println(partial1)

    val partial1_3 = partial1(3)
    assert(partial1_3 == 4)

    assert(f(4)(5) == 9)
  }

  val f: Int => Int => Int = x => y => x + y

  def g(x: Int, y: Int) = x + y
}

File: java/src/P04PartialApplication.java
import java.util.function.Function;

public class P04PartialApplication {
    public static void main(String[] args) {
        var partial1 = f.apply(1);
        System.out.println(partial1.getClass());
        System.out.println(partial1);

        var partial1_3 = partial1.apply(3);
        assert 4 == partial1_3;

        assert 9 == f.apply(4).apply(5);
    }

    private final static Function<Integer, Function<Integer, Integer>> f = x -> y -> x + y;

    private static Integer g(Integer x, Integer y) {
        return x + y;
    }
}